runtime/tests/checker/runtimetype_test.go

```diff
@@ -0,0 +1,772 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2019-2021 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package checker
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/cadence/runtime/sema"
+)
+
+func TestCheckOptionalTypeConstructor(t *testing.T) {
+
+	t.Parallel()
+
+	cases := []struct {
+		name  string
+		code  string
+		valid bool
+	}{
+		{
+			name: "String",
+			code: `
+              let result = OptionalType(Type<String>())
+            `,
+			valid: true,
+		},
+		{
+			name: "Int",
+			code: `
+				let result = OptionalType(Type<Int>())
+            `,
+			valid: true,
+		},
+		{
+			name: "resource",
+			code: `
+              resource R {}
+              let result = OptionalType(Type<@R>())
+            `,
+			valid: true,
+		},
+		{
+			name: "type mismatch",
+			code: `
+              let result = OptionalType(3)
+            `,
+			valid: false,
+		},
+		{
+			name: "too many args",
+			code: `
+              let result = OptionalType(Type<Int>(), Type<Int>())
+            `,
+			valid: false,
+		},
+		{
+			name: "too few args",
+			code: `
+              let result = OptionalType()
+            `,
+			valid: false,
+		},
+	}
+
+	for _, testCase := range cases {
+		t.Run(testCase.name, func(t *testing.T) {
+			checker, err := ParseAndCheck(t, testCase.code)
+
+			if testCase.valid {
+				require.NoError(t, err)
+				assert.Equal(t,
+					sema.MetaType,
+					RequireGlobalValue(t, checker.Elaboration, "result"),
+				)
+			} else {
+				require.Error(t, err)
```

Comment:
Maybe assert the expected errors where reported, e.g by changing the `valid bool` field of the test case to `expectedErrors []error` and using `ExpectCheckerErrors` . Same for the other test functions below